crates/accelerate/src/sparse_observable.rs

```diff
@@ -0,0 +1,1677 @@
+/// matrices.  In this analogy, the terms of the sum are the "rows", and the qubit terms are the
+/// "columns", where an absent entry represents the identity rather than a zero.  More explicitly,
+/// the representation is made up of four contiguous arrays:
+///
+/// .. _sparse-observable-arrays:
+/// .. table:: Data arrays used to represent :class:`.SparseObservable`
+///
+///   ==================  ===========  =============================================================
+///   Attribute           Length       Description
+///   ==================  ===========  =============================================================
+///   :attr:`coeffs`      :math:`t`    The complex scalar multiplier for each term.
+///
+///   :attr:`bit_terms`   :math:`s`    Each of the non-identity single-qubit terms for all of the
+///                                    operators, in order.  These correspond to the non-identity
+///                                    :math:`A^{(n)}_i` in the sum description, where the entries
+///                                    are stored in order of increasing :math:`i` first, and in
+///                                    order of increasing :math:`n` within each term.
+///
+///   :attr:`indices`     :math:`s`    The corresponding qubit (:math:`n`) for each of the operators
+///                                    in :attr:`bit_terms`.  :class:`SparseObservable` requires
+///                                    that this list is term-wise sorted, and algorithms can rely
+///                                    on this invariant being upheld.
+///
+///   :attr:`boundaries`  :math:`t+1`  The indices that partition :attr:`bit_terms` and
+///                                    :attr:`indices` into complete terms.  For term number
+///                                    :math:`i`, its complex coefficient is ``coeffs[i]``, and its
+///                                    non-identity single-qubit operators and their corresponding
+///                                    qubits are the slice ``boundaries[i] : boundaries[i+1]`` into
+///                                    :attr:`bit_terms` and :attr:`indices` respectively.
+///                                    :attr:`boundaries` always has an explicit 0 as its first
+///                                    element.
+///   ==================  ===========  =============================================================
+///
+/// The length parameter :math:`t` is the number of terms in the sum, and the parameter :math:`s` is
+/// the total number of non-identity single-qubit terms.
+///
+/// As illustrative examples:
+///
+/// * in the case of a zero operator, :attr:`boundaries` is length 1 (a single 0) and all other
+///   vectors are empty.
+/// * in the case of a fully simplified identity operator, :attr:`boundaries` is ``[0, 0]``,
+///   :attr:`coeffs` has a single entry, and :attr:`bit_terms` and :attr:`indices` are empty.
```

Comment:
A simple (but non-trivial 😉 ) example might be nice. This is partially for my own understanding; `s` includes all single-qubit Paulis, right? (I.e. `ZZ` would have `s=2`?)
```suggestion
///   :attr:`coeffs` has a single entry, and :attr:`bit_terms` and :attr:`indices` are empty.
/// * a ``ZZ - IX`` observable would have :math:`t=2` coefficients (namely ``1`` and ``-1``)
///   and :math:`s=3` single-qubit Pauli terms (namely ``Z, Z, X``)
```

Reply:
Done in 9dd0187213.